Add interpreter support for Handles/PythonOp/CppOp

zdevito opened the pull request:
This treats Handles as a first-class type in the interpreter
since this turned out to be conceptually simpler than treating
them as a separate concept, which requires a second channel for
register allocating and moving data from one op to the next.

Notes:
* The refcounting nature of tensors is factored into its own base type
so that it can be shared with other refcounted types such as handle.
* Some methods redundant with TensorBase have been deleted from Tensor
* The interpreter uses raw refcounted handles. In addition to being
able to treat Tensors and Handles as the same base object, it removes
a lot of redundant refcounting as objects moved from tensors to input/
output lists.
* aten_dispatch has been updated to work directly on the raw refcounted
lists to avoid refcounting and duplicate lists.
* PythonOp and CppOp gain VariableFlags lists so that the interpreter knows how the bodies of those operators should be executed.

zdevito committed:
Add interpreter support for Handles/PythonOp/CppOp

This treats Handles as a first-class type in the interpreter
since this turned out to be conceptually simpler than treating
them as a separate concept, which requires a second channel for
register allocating and moving data from one op to the next.

Notes:
* The refcounting nature of tensors is factored into its own base type
so that it can be shared with other refcounted types such as handle.
* Some methods redundant with TensorBase have been deleted from Tensor
* The interpreter uses raw refcounted handles. In addition to being
able to treat Tensors and Handles as the same base object, it removes
a lot of redundant refcounting as objects moved from tensors to input/
output lists.
* aten_dispatch has been updated to work directly on the raw refcounted
lists to avoid refcounting and duplicate lists.

zdevito commented:
@apaszke this currently is failing in test_backward_closure, becuase I think this uncovered something the InterpreterAutogradFunction that appears to be due to Edge pruning semantics. Not sure what is wrong but I figured I'd put the code up here so you might know what is wrong.

```
======================================================================
ERROR: test_backward_closure (__main__.TestJit)
Check that autograd closures handle multiple stages correctly.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test/test_jit.py", line 506, in test_backward_closure
    grad_x.backward()
  File "/data/users/zdevito/pytorch/torch/autograd/variable.py", line 128, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph, retain_variables)
  File "/data/users/zdevito/pytorch/torch/autograd/__init__.py", line 98, in backward
    variables, grad_variables, retain_graph)
RuntimeError: Function 'N5torch3jit27InterpreterAutogradFunctionE' returned an invalid number of outputs - expected 0, but got 1
```

----------------------------------------------------------------------

apaszke commented:
Nothing comes to my mind immediately. I'd need to see the full graph. In general, it might mean that there's something wrong with the way `next_functions` are filled in (i.e. some are missing).

zdevito committed:
Removing jit_closure.cpp

The interpreter can now handle all pathways.

zdevito committed:
Fix a bug in how copied_next_functions are added in the interpreter.

ezyang commented on tools/jit/templates/aten_dispatch.cpp:
```diff
@@ -20,19 +20,53 @@ using operator_constructor = std::function<TensorOp(jit::Node*)>;
-  outputs.insert(outputs.end(), t.begin(), t.end());
+// a temporary Tensor that does not alter the refcount of impl on
+// acquisition or release, avoids any refcounting in dispatch functions
+struct TensorTemporary : public at::Tensor {
```
Is this https://github.com/zdevito/ATen/issues/114 ?

zdevito replied:
Yes, but I don't want to expose it in aten's header files because it will be very confusing to have two Tensor types, so it is just created locally in the dispatch implementation.

ezyang commented on tools/jit/templates/aten_dispatch.cpp:
```diff
@@ -20,19 +20,53 @@ using operator_constructor = std::function<TensorOp(jit::Node*)>;
+};
+
+// same thing but creates a tensor list, only used rarely (e.g. for cat)
+struct TensorTemporaryList {
```
I had to do some mental work to convince myself this was safe, it might be worth writing down: the key is that a `TensorTemporaryList` only exposes a `TensorList`, not its underlying `std::vector<at::Tensor>`. This ArrayRef has the desired semantics: if you get out an `at::Tensor` from it, the refcount is bumped; if you take a reference, it is only guaranteed to stay live as long as the ArrayRef is live, which is exactly the guarantee you need to use this safely.

ezyang commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -1,42 +1,208 @@
 #include "torch/csrc/jit/ir.h"
 #include "torch/csrc/autograd/profiler.h"
 #include "torch/csrc/jit/generated/aten_dispatch.h"
+#include "Python.h"
```
Second time you included it!

apaszke commented on tools/jit/templates/aten_dispatch.cpp:
```diff
@@ -20,19 +20,53 @@ using operator_constructor = std::function<TensorOp(jit::Node*)>;
+};
+
+using refcounted_list = std::vector<at::RefCounted*>;
+void pack_list(refcounted_list & outputs, Tensor v) { outputs.push_back(v.detach()); }
```
Tensors to pack are never `TemporaryTensor`s right? Can you add a comment that this handle mechanism is for inputs only?

zdevito replied:
They are never temporaries and the class TensorTemporaries is implemented such that if you tried, it would copy the reference into a non-temporary tensor first.

apaszke commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -1,42 +1,208 @@
 #include "torch/csrc/autograd/profiler.h"
 #include "torch/csrc/jit/generated/aten_dispatch.h"
+#include "Python.h"
+#include "pybind11/pybind11.h"
```
I think it's a better idea to include `torch/csrc/jit/pybind.h`. It already defines some useful wrappers

apaszke commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -1,42 +1,208 @@
+};
+
+struct Handle : at::RefCounted {
+  std::shared_ptr<DummyFunction> forward_inputs;
```
Can you add a comment here? I can't see how a "DummyFunction" specifies a list of forward inputs. The way you use it later makes sense, but that's not immediately obvious when you start reading the file.

apaszke commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -1,42 +1,208 @@
+};
+
+at::Tensor unsafeTensorBorrow(at::RefCounted * rc) {
+  return at::Tensor(static_cast<at::TensorImpl*>(rc), true);
```
How is that a borrow? You're retaining the impl, right? It's just an `unsafeTensorCast`

apaszke commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -1,42 +1,208 @@
+    } else {
+      autograd::VarFlags flags = {false, false};
+      return autograd::make_variable(unsafeTensorBorrow(input), flags);
+    }
```
You're not handling volatile (which is not strictly necessary for correctness, but might be nice to have). Can you at least leave a note saying that it's on purpose or that it's a todo?

apaszke commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -1,42 +1,208 @@
+  }
+  autograd::Variable addInput(at::RefCounted* input, const VariableFlags & flags_) {
+    if(handle && flags_.requires_grad) {
+      autograd::VarFlags flags = {flags_.requires_grad, flags_.is_volatile};
```
Well these flags are guaranteed to be `{true, false}` here

apaszke commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -1,42 +1,208 @@
+Operation createPythonCallback(PythonOp* op) {
+  py::object func = py::handle(op->pyobj.get()).attr("apply");
+  bool has_handle = hasHandleOutput(op);
+  return [=](const refcounted_list & inputs, refcounted_list & outputs) {
```
This implementation adds an important constraint that wasn't true before - we have to retain the JIT graph even after it was compiled, because you need the `op` pointer at run time! We should document this somewhere (probably in `python_compiled_function.cpp`, next to the `graph_` field).

zdevito replied:
Added docs in jit::Code to remind people not to remove the graph handle.

apaszke commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -1,42 +1,208 @@
+      } else if(arg_type == 't') {
+        py_inputs[i] = THPVariable_Wrap(
+          builder.addInput(inputs.at(next_tensor), op->var_flags.at(next_tensor)));
+        next_tensor++;
```
Can you `JIT_ASSERT` that we have enough tensor inputs?

apaszke commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -1,42 +1,208 @@
+    if(handle && flags_.requires_grad) {
+      autograd::VarFlags flags = {flags_.requires_grad, flags_.is_volatile};
+      return autograd::make_variable(
+        unsafeTensorBorrow(input),
```
I'm pretty sure that you're wrapping Variables into Variables here. I don't think we unwrap the Variables before we put them in the interpreter anywhere

zdevito replied:
This line already exists in `InterpreterAutogradFunction::apply`, so we are in fact working on Tensors already.
 
```
auto tinputs = fmap(inputs, [](const autograd::Variable& i) { return i.data(); });
```

apaszke commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -1,42 +1,208 @@
+    for(auto & output : v_outputs) {
+      outputs.push_back(builder.addOutput(output));
+    }
+    builder.writeTo(outputs);
```
It's a bit weird to use HandleBuilder even when you're not building handles, just because it can write to outputs that have nothing to do with the actual handle.

zdevito replied:
I added a comment about this. Handles interact with how outputs/inputs are converted so it makes sense to put it one place.

apaszke commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -1,42 +1,208 @@
+      return false; // stop output and do not run DummyFunction
+    });
+    // node handle_in->use_count() == 1 means that we are guarenteed that we have the only
+    // only copy of the backward pass, but it is not clear that it is safe even in that case
```
"it is not clear that it is safe even in that case" - can you make this more specific? I also can't see how `use_count() == 1` implies that we have "the only copy of backward pass" (especially that I don't know what a copy of backward pass is).

zdevito replied:
So you have something like:

stage 1: handle0 -> Function0 -> Function1 -> Function2 -> Grad Accumulator
stage 2: handle1 -> Function3 -> Function 4 -> Function0 -> Grad Accumulator

Function 0 exists in both graphs because copied_next_fns told handle1 to inherit it from handle0 to calculate a derivative path that goes through the earlier stage.

Even if handle1.use_count() == 1, you can't free its graph unless handle0->use_count() == 0.
I think the 'right' solution here is for 'keep_graph=False' to actually mean "release the variables incrementally as this is run if we are guaranteed that the Function will be freed after the autograd finishes". But that condition is hard to check so is probably not worth trying to do.

apaszke commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -1,42 +1,208 @@
+    // node handle_in->use_count() == 1 means that we are guarenteed that we have the only
+    // only copy of the backward pass, but it is not clear that it is safe even in that case
+    // if use_count() > 1 then it is possible it drops to 1 during this execute,
+    // however, that is safe if conservative since it will get freed when the handles die.
```
I also don't know how could this happen. Please expand these comments

apaszke commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -1,42 +1,208 @@
+      for(auto & i : inputs) {
+        tinputs.push_back(unsafeTensorBorrow(i));
+      }
+      fusion_fn->launch(tinputs, toutputs);
```
This could accept those temporary lists right?

zdevito replied:
Not easily -- You _cannot_ pass a TensorTemporary to a thing that takes std::vector<Tensor> because it will slice off the Temporary and then (incorrectly) delete the tensor. I am going to change around the implementation of TensorTemporary to make its use safer.

apaszke commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -108,6 +275,7 @@ struct CodeImpl {
     // so we clean it up
     // this is done with a backward scan where we mark the first time we see it
     std::unordered_set<int> seen_registers;
+    std::unordered_set<int> seen_handles;
```
is this even used?

zdevito replied:
nope

apaszke commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -193,6 +362,55 @@ struct CodeImpl {
+  : registers(rhs.registers) {
+    for(auto & r : registers) {
+      if(isValid(r))
+        r->retain();
```
Doesn't `refcounted_list` automatically bump the refcount when you copy it?

zdevito replied:
I renamed it list_of_refcounted because refcounted_list is highly ambiguous.

apaszke commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -193,6 +362,55 @@ struct CodeImpl {
+  at::RefCounted*& operator[](size_t i) {
+    return registers[i];
+  }
+  //guarenteed to be non-null
```
What is guaranteed to be non-null? You have a `isValid` check inside, so it seems like it could be null.

zdevito replied:
(no longer exists)

apaszke commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -193,6 +362,55 @@ struct CodeImpl {
+  at::Tensor load(size_t i, bool andFree) {
+    auto & v = registers[i];
+    if(andFree) {
+      auto r = at::Tensor(static_cast<at::TensorImpl*>(v), false);
```
Now, that's an `unsafeTensorBorrow` ;)

zdevito replied:
per pybind11:

```
at::Tensor(static_cast<at::TensorImpl*>(v), false) <-- reinterpret_steal
at::Tensor(static_cast<at::TensorImpl*>(v), true) <-- reinterpret_borrow
```

Pybind11's terminology however, is confusing, so I am going to change the names and just be more explicit.

apaszke commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -193,6 +362,55 @@ struct CodeImpl {
+      v = nullptr;
+      return r;
+    } else {
+      return at::Tensor(static_cast<at::TensorImpl*>(v), true);
```
nit: you can use that tensor caster you defined earlier

apaszke commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -204,24 +422,35 @@ struct InterpreterStateImpl {
           int reg = get(inst.outputs,i);
-          registers[reg] = std::move(output_buffer[i]);
-          //std::cout << "registers[" << reg << "] = outputs[" << i << "](" << output_buffer[i].defined() << ")\n";
+          registers[reg] = output_buffer[i];
```
Can't you use `.save` here to steal the references?

---

Now it seems to me that this line is leaking (since `refcounted_list` doesn't do refcounting, and you don't steal the references, `output_buffer` won't `decref` the things it holds).

zdevito replied:
This is fine - ownership is being transferred from `output_buffer` to `register`. output_buffer won't release, but registers also wont retain. It will be clearer with better types, which I will add.

apaszke commented on tools/jit/templates/aten_dispatch.cpp:
```diff
@@ -20,19 +20,53 @@ using operator_constructor = std::function<TensorOp(jit::Node*)>;
+  std::vector<at::Tensor> tensors;
+};
+
+using refcounted_list = std::vector<at::RefCounted*>;
```
Something that was very surprising for me later in the code is that `refcounted_list` doesn't increment the refcount when it's copied or decrement when it's destroyed! I'd say that the name implies the contrary.

apaszke commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -193,6 +362,55 @@ struct CodeImpl {
+        r->release();
+    }
+  }
+  at::RefCounted*& operator[](size_t i) {
```
Returning a reference in here seems dangerous - if you have a managed object in `i`th position, and you just overwrite its pointer with another one, then the previous one won't be freed.

At least describing the intended semantics would be helpful.

zdevito replied:
Ill try to refactor so that we have owning/non-owning implementations.

apaszke reviewed (commented):
The code looks good for the most part. I only have two important issues with this PR.

One is that the handle mechanism is new and completely undocumented, so it would be nice to add a comment at the top of interpreter, explaining how they work now (this would have made the review easier too). 

The other, and more important, is ref counting. Did you see any perf improvements in the benchmark thanks to going around it? The code is really complicated now, because it's operating on ref counted types, but sometimes completely ignoring the ref counts for perf reasons, and the size of it makes it very hard for me to reason about leaks. I'm not sure that this ref counting scheme is correct at all. A comment that clearly outlines where do we ignore the conventions and do our things would be very helpful for this. Additionally, every object that deals with custom ref counting should have steal/borrow semantics stated next to every method.

zdevito committed:
Minor PR comments

zdevito committed:
Codemod RefCounted -> Retainable

zdevito committed:
Add ownership tracking functions for Retainable

These functions like `unsafeToTensorShare` describe how
ownership transfers in the interpreter. The `Steal` variants
take rvalue references as arguments, and invalidate those
arguments to prevent potential problems.

Registers is replace with a Retainable list that is focused
solely on ownership to make it more clear.

zdevito committed:
Make TensorTemporary safer

This removes the subtype relationship because it is too easy to
do something horribly unsafe:

```
  void foo(at::Tensor bar) {
    // bar destructor call release on a temporary!
  }

  foo(TensorTemporary(retainable)); // structure slicing!
```

The new code is basically the same but does not use a subtype so
that structure slicing cannot happen.

zdevito commented:
This is now updated with PR comments a few notes:

* RefCounted was renamed to Retainable because refcounted_list is grammatically ambiguous.
* TensorTemporary is no longer a subclass of Tensor, because it can lead to unsafe code when structure slicing occurs. It is now its own object that manages the lifetime of a special temporary tensor.
* all places where Retainables move around are now done through functions that describe the ownership transfer semantics taking place.

apaszke commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -1,42 +1,229 @@
+    for(size_t i = 0; i < inputs.size(); i++) {
+      v_inputs.push_back(builder.addInput(inputs[i], op->var_flags[i]));
+    }
+    autograd::variable_list v_outputs = func->apply(v_inputs);
```
This should do `(*func)(v_inputs)`

apaszke commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -1,42 +1,229 @@
+    // note: node handle_in->use_count() == 1 means that we are guarenteed that we have the only
+    // only copy of the handle. This might make it seems like we can pass keep_graph=False.
+    // However, it is possible for 'copied_next_fns' to grab functions used by _other_ handles,
+    // so it is not to dispose of the graph.
```
I still don't understand this comment, but at least the code is safe now 😄

apaszke commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -204,24 +448,35 @@ struct InterpreterStateImpl {
+        auto & inputs = inst.inputs.values;
+        for(int i = 0; i < inputs.size; i++) {
+          int reg = get(inputs,i);
+          input_buffer.push_back(registers[reg]);
```
So this effectively creates a temporary, because it only copies pointers to retainables, correct?

zdevito replied:
Yes, the input/output buffers are just short term places where inputs/outputs get stored before moving into registers. The output list is unmanaged, but during the time it is unempty the only things that happen are appending to the output list and writting the outputs back into registers.

apaszke commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -266,12 +519,12 @@ struct InterpreterStateImpl {
 
   // single buffer for input calls to ATen functions, so that we do not reallocate
-  std::vector<at::Tensor> input_buffer;
+  std::vector<at::Retainable*> input_buffer;
```
isn't this `list_of_retainable`? :wink:

apaszke commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -238,15 +493,13 @@ struct InterpreterStateImpl {
   void loadTensorsFromRegisters(const UseList & uses, std::vector<at::Tensor> & outputs) {
```
BTW this is only used once now, for gathering outputs.

apaszke commented on tools/jit/templates/aten_dispatch.cpp:
```diff
@@ -20,19 +20,75 @@ using operator_constructor = std::function<TensorOp(jit::Node*)>;
+// it has a owning reference and then that reference is stolen ad added to the
+// list_of_retainable output list.
+// pack_list never operates on tensor temporaries.
+void pack_list(list_of_retainable & outputs, Tensor v) { outputs.push_back(toRetainableSteal(std::move(v))); }
```
It's still a bit ugly that `list_of_retainable` doesn't manage the lifetime of the pointers it holds - some code might look as if it should free memory when an error happens, while in reality you have these unsafe moments before you plug these pointers back to registers, when any exceptions will basically leak memory.

Can't we just have an `std::vector<at::Retainable*>` for inputs, because they are temporaries that don't bump refcounts, and `std::vector<std::unique_ptr<at::Retainable>>` (with a custom deleter) to ensure we will never leak outputs?

zdevito replied:
I tried using std::vectors of unique_ptrs for the register list and for outputs. It doesn't really work: for outputs, you add a whole useless loop during the destruction of the list to check for destructors that will never run. for the register list, it can't be a trivial vector<unique_ptr> list because it has to support a copy constructor. It can be a new container that uses a unique_ptr underneath, but that container is almost as long as the implementation of the current container anyway with little added benefit. Working around this stuff just doesn't seem worth the effort.

apaszke replied:
Ok, let's leave it for later :+1:

apaszke commented on tools/jit/templates/aten_dispatch.h:
```diff
@@ -7,15 +7,52 @@
-  TensorOp(op_type op, std::string name, size_t num_inputs)
+// a list of refcounted pointers. this is not called refcounted_list because
+// that sounds like a refcounted list of pointers, which is something else.
+using list_of_retainable = std::vector<at::Retainable*>;
```
I think `temp_retainable_list` would be pretty clear. `list_of_retainable` doesn't imply that they won't be freed when it's destructed.

apaszke reviewed (commented):
This looks really good. I think we could still improve the safety by using a different type for outputs vector (there's no reason why it has to be of the same type as `inputs`), but that's pretty much it.

apaszke commented:
And this `std::unique_ptr<at::Retainable>` could also be used to implement the struct we use to hold registers, without the need to iterate through these vectors and checking if we don't need to free sth. It will do it automatically when you destruct it, or reset to a `nullptr`.

zdevito committed:
lint

zdevito commented:
I fixed the minor stuff. I tried using unique_ptrs in both registers and the output lists but as noted before they each have issues. I'd like to move on to other things, so I am going to merge this assuming there aren't major issues.

apaszke commented:
I think this is good to go, but I'd be great if you could answer a few questions I posted during the last review. Also, you need to rebase it on top of master because it conflicts with Sam's changes.

zdevito committed:
Update to fix rebase build errors.
